### PR TITLE
Start to use the common condition infrastructure from knative/pkg.

### PR DESCRIPTION
**Pull request by mattmoor**
WIP until https://github.com/knative/build/pull/361 merges.

cc @n3wscott

### CODE REVIEW

File: pkg/apis/build/v1alpha1/build_types.go [hunk 1]
```diff
@@ -209,12 +214,17 @@ type BuildStatus struct {
...
 	// StepStates describes the state of each build step container.
 	StepStates []corev1.ContainerState `json:"stepStates,omitEmpty"`
-	// Conditions describes the set of conditions of this build.
-	Conditions []BuildCondition `json:"conditions,omitempty"`
+
 	// StepsCompleted lists the name of build steps completed.
 	StepsCompleted []string `json:"stepsCompleted"`
+
+	// Conditions describes the set of conditions of this build.
+	Conditions duckv1alpha1.Conditions `json:"conditions,omitempty"`
```
Comment (n3wscott) [inline pkg/apis/build/v1alpha1/build_types.go]: you can update this to the following:

```
Conditions duck.Conditions `json:"conditions,omitempty" patchStrategy:"merge" patchMergeKey:"type"`
```

File: pkg/apis/build/v1alpha1/build_types.go [hunk 2]
```diff
@@ -271,42 +260,28 @@ type BuildList struct {
...
 // GetCondition returns the Condition matching the given type.
-func (bs *BuildStatus) GetCondition(t BuildConditionType) *BuildCondition {
-	for _, cond := range bs.Conditions {
-		if cond.Type == t {
-			return &cond
-		}
-	}
-	return nil
+func (bs *BuildStatus) GetCondition(t duckv1alpha1.ConditionType) *duckv1alpha1.Condition {
+	return buildCondSet.Manage(bs).GetCondition(t)
```
Comment (n3wscott) [inline pkg/apis/build/v1alpha1/build_types.go]: LOL `manage bs`

File: pkg/apis/build/v1alpha1/build_types.go [hunk 3]
```diff
@@ -271,42 +260,28 @@ type BuildList struct {
...
 }
 
 // SetCondition sets the condition, unsetting previous conditions with the same
 // type as necessary.
-func (b *BuildStatus) SetCondition(newCond *BuildCondition) {
-	if newCond == nil {
-		return
+func (bs *BuildStatus) SetCondition(newCond *duckv1alpha1.Condition) {
```
Comment (n3wscott) [inline pkg/apis/build/v1alpha1/build_types.go]: you should be able to remove this entirely and replace it with `Mark*` methods. Looks like there is only Succeeded and Failed. I would implement:

`MarkBuildSucceded`
`MarkBuildFailed`

and then remove SetCondition entirely.

File: pkg/builder/cluster/builder.go [hunk 4]
```diff
@@ -75,7 +76,7 @@ func (op *operation) Checkpoint(build *v1alpha1.Build, status *v1alpha1.BuildSta
 		}
 		status.StepStates = append(status.StepStates, s.State)
 	}
-	status.SetCondition(&v1alpha1.BuildCondition{
```
Comment (n3wscott) [inline pkg/builder/cluster/builder.go]: this should be `status.Initialize()`

File: pkg/builder/cluster/builder.go [hunk 5]
```diff
@@ -134,21 +135,21 @@ func (op *operation) Wait() (*v1alpha1.BuildStatus, error) {
 
 	if pod.Status.Phase == corev1.PodFailed {
 		msg := getFailureMessage(pod)
-		bs.SetCondition(&v1alpha1.BuildCondition{
```
Comment (n3wscott) [inline pkg/builder/cluster/builder.go]: And then use the correct `Mark` method

File: pkg/apis/build/v1alpha1/build_types.go (hunk 3 above)
Comment (mattmoor) [reply pkg/apis/build/v1alpha1/build_types.go]: Ack.  Build hasn't yet adopted this model, so I think that should be a follow up.

**Comment by mattmoor** [comment]
/hold

There is something going on here that I need to look into.

**Comment by mattmoor** [comment]
/hold cancel

I had to update the e2e script since apparently the serialized field name changed.

File: pkg/apis/build/v1alpha1/build_types.go (hunk 1 above)
Comment (mattmoor) [reply pkg/apis/build/v1alpha1/build_types.go]: This is only going to be meaningful for API servers.

**Pull request merged**